api/workload/v2/reader.go

```diff
@@ -0,0 +1,94 @@
+package workload
+
+import (
+	"context"
+	"sync"
+
+	"github.com/sirupsen/logrus"
+	"github.com/spiffe/spire/proto/api/workload"
+)
+
+type streamReader struct {
+	Chan          chan *workload.X509SVIDResponse
+	logger        *logrus.Logger
+	streamManager *streamManager
+
+	// atomic stream
+	mu     *sync.RWMutex
+	stream *managedStream
+}
+
+func newStreamReader(ctx context.Context, logger *logrus.Logger, streamManager *streamManager) *streamReader {
+	r := &streamReader{
+		logger:        logger,
+		streamManager: streamManager,
+		mu:            new(sync.RWMutex),
+		Chan:          make(chan *workload.X509SVIDResponse),
+	}
+	r.start(ctx)
+	return r
+}
+
+func (c *streamReader) getStream() *managedStream {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+	return c.stream
+}
+
+func (c *streamReader) setStream(stream *managedStream) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	c.stream = stream
+}
+
+func (c *streamReader) start(ctx context.Context) {
+	c.logger.Debug("Starting reader.")
+	go func() {
+		defer c.logger.Debug("Shutting down reader")
+		defer c.closeStream()
+		defer close(c.Chan)
+
+		for {
+
+			select {
+			case stream, ok := <-c.streamManager.Chan:
```

Comment:
I think this hunk needs to move into `Client` so that the reader can be more focused on the lifecycle of a *single* stream.

Reply:
I could see this both ways.. if a reader is only focused on a single stream, then the client must be aware of the relationship between streams and readers, when perhaps the client should only be concerned with the receiving from the reader itself, and unaware of the relationship between a reader and its stream. that is: "the client knows that the reader needs a stream manager" rather than "the client knows that when the stream is interrupted, it must create a new reader"

maybe let's chat 1:1 about it

Reply:
I think it's okay for the client to know about these details since they are all internal to it. I find nested infinite for loops a pain to read, and sometimes to test. I'm also fine either way but you may find yourself writing complicated logic into the client test harness because you are trying to test these reader details via an integration test. However, the complexity is going to exist somewhere.